### DISCUSSION TITLE
@types/node exposes __dirname __filename and require even when using ESM

### DISCUSSION THREAD

**Discussion by Exifers** [DefinitelyTyped/DefinitelyTyped#66780] [_Misc]
Currently @types/node exposes globals that are supposed to be used only with CJS modules. When using ESM modules, this means we have autocompletion for them, and Typescript will compile fine.

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/24790356/d867feff-ba8f-44e5-99a5-b52fbb539ce2)

At runtime though, it fails if we use them `__dirname is not defined in ES module scope`.

Is there any way we could have Node type definitions without these CJS globals?
Would it make sense to have multiple entrypoints to this package, like @types/node/esm and @types/node/cjs, to separate ESM-only and CJS-only types?

**Comment by ayame113** [comment]
I agree...if we are using ESM, `require`, `exports`, `module`, `__filename`, `__dirname`, `global` should not be exposed as global variables.

I think the `exports` field in package.json might be useful to solve this problem.
https://www.typescriptlang.org/docs/handbook/esm-node.html#:~:text=types%22%20import%20condition.-,//%20package.json,-%7B

Issues that seem related: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42201

**Comment by aphex** [comment]
Just ran into this as well. It does seem strange there is no way to swap the type from `cjs` to `esm`. Looking at `exports` from `package.json`. Does it provide a way to export different type files depending on `require` vs `import`? I am only seeing code, but types seem to be common.

As mentioned there may just need to be multiple entry points. For a quick hack fix you can `pnpm patch` the package and remove the lines from `globals.d.ts`

```diff
diff --git a/globals.d.ts b/globals.d.ts
index 5f250062b0295a95916ad2acd6ae8141ff2b5e71..35f4cdd3fd6f6c6dd5f2f51ae5bbe93430695ea2 100644
--- a/globals.d.ts
+++ b/globals.d.ts
@@ -44,15 +44,6 @@ declare global {
     var process: NodeJS.Process;
     var console: Console;
 
-    var __filename: string;
-    var __dirname: string;
-
-    var require: NodeRequire;
-    var module: NodeModule;
-
-    // Same as module.exports
-    var exports: any;
-
     /**
      * Only available if `--expose-gc` is passed to the process.
      */
```
